File: 20. Add two numbers represented by lists.py
class Node:
    def __init__(self,data): self.data=data; self.next=None

def add_lists(a,b):
    head=None; tail=None; carry=0
    while a or b or carry:
        s=carry
        if a:
            s+=a.data; a=a.next
        if b:
            s+=b.data; b=b.next
        node=Node(s%10)
        carry=s//10
        if not head:
            head=tail=node
        else:
            tail.next=node; tail=node
    return head

def print_list(h):
    cur=h; out=[]
    while cur:
        out.append(str(cur.data)); cur=cur.next
    print(' -> '.join(out))

if __name__=='__main__':
    a=Node(2); a.next=Node(4); a.next.next=Node(3)
    b=Node(5); b.next=Node(6); b.next.next=Node(4)
    s=add_lists(a,b)
    print_list(s)

File: 20. Add two numbers represented by lists.cpp
#include <iostream>
using namespace std; struct Node{int data; Node* next; Node(int d):data(d),next(nullptr){}
};
Node* addLists(Node* a, Node* b){ Node* head=nullptr; Node* tail=nullptr; int carry=0; while(a||b||carry){ int s=carry; if(a){ s+=a->data; a=a->next;} if(b){ s+=b->data; b=b->next;} Node* n=new Node(s%10); carry=s/10; if(!head) head=tail=n; else { tail->next=n; tail=n; } } return head; }
void print(Node* h){ while(h){ cout<<h->data; if(h->next) cout<<" -> "; h=h->next;} cout<<endl; }
int main(){ Node* a=new Node(2); a->next=new Node(4); a->next->next=new Node(3); Node* b=new Node(5); b->next=new Node(6); b->next->next=new Node(4); Node* s=addLists(a,b); print(s); return 0; }
